### PR TITLE
Feature/translate json functions #29306

### LINKED ISSUE #29306: Provide EF.Functions translations for JSON functions
For users using JSON without a strongly-typed schema, mapping the column as a string is the only option at the moment. We should provide translations for JSON_VALUE/JSON_QUERY for SQL Server and json_extract for SQLite.

### PR DESCRIPTION
**Pull request by joelmandell**
Fixes #29306

- [x] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)
- [x] I've posted a comment on an issue with a detailed description of how I am planning to contribute and got approval from a member of the team
- [x] The code builds and tests pass locally (also verified by our automated build checks)
- [ ] Commit messages follow this format: (**This I totally forgot**)
```
        Summary of the changes
        - Detail 1
        - Detail 2

        Fixes #bugnumber
```
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code follows the same patterns and style as existing code in this repo (**Kinda... little bit different with unit-tests. But want to get your feedback if that's okay)**.

### CODE REVIEW

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 1]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
 /// </remarks>
 public static class SqlServerDbFunctionsExtensions
 {
+    /// <summary>
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:19]: Please move these to the bottom, after the other existing functions.

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 2]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
...
+    /// </summary>
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-database-functions">Database functions</see>, and
+    ///     <see href="https://aka.ms/efcore-docs-sqlserver">Accessing SQL Server and SQL Azure databases with EF Core</see>
+    ///     for more information and examples.
+    /// </remarks>
+    /// <param name="_">The <see cref="DbFunctions" /> instance.</param>
+    /// <param name="propertyReference">The property in table</param>
+    /// <param name="path">The json path.</param>
+    /// <returns>Value of path</returns>
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:31]: Copied reduced versions of the descriptions from the SQL Server docs, and added a link to them.
```suggestion
    /// <param name="expression">An expression. Typically the name of a variable or a column that contains JSON text.</param>
    /// <param name="path">A JSON path that specifies the property to extract.</param>
    /// <returns>A single text value.</returns>
    /// <seealso href="https://learn.microsoft.com/sql/t-sql/functions/json-value-transact-sql">SQL Server documentation for <c>JSON_VALUE</c>.</seealso>
```

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 3]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
...
+    public static string JsonValue(
+        this DbFunctions _,
+        object propertyReference,
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:34]: JSON_VALUE doesn't have to accept a column, it can be any string expression that evaluates to a JSON string (here again, am following the SQL Server docs).

```suggestion
        object expression,
```

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 4]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
 /// </remarks>
 public static class SqlServerDbFunctionsExtensions
 {
+    /// <summary>
+    ///     Extracts a scalar value from a JSON string. <paramref name="propertyReference" /> and <paramref name="path" />.
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:20]:
```suggestion
    ///     Extracts a scalar value from a JSON string.
```

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 5]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
...
+    /// <returns>Value of path</returns>
+    public static string JsonValue(
+        this DbFunctions _,
+        object propertyReference,
+        string path)
+        => throw new InvalidOperationException(CoreStrings.FunctionOnClient(nameof(SqlServerDbFunctionsExtensions.JsonValue)));
+
+    /// <summary>
+    ///     Extracts an object or an array from a JSON string. <paramref name="propertyReference" /> and <paramref name="path" />.
+    ///     Corresponds to SQL Server's <c>JSON_VALUE(@propertyReference, @path)</c>.
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:40]:
```suggestion
    ///     Extracts an object or an array from a JSON string.
    ///     Corresponds to SQL Server's <c>JSON_QUERY(@expression, @path)</c>.
```

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 6]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
...
+    /// </summary>
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-database-functions">Database functions</see>, and
+    ///     <see href="https://aka.ms/efcore-docs-sqlserver">Accessing SQL Server and SQL Azure databases with EF Core</see>
+    ///     for more information and examples.
+    /// </remarks>
+    /// <param name="_">The <see cref="DbFunctions" /> instance.</param>
+    /// <param name="propertyReference">The property in table</param>
+    /// <param name="path">The json path.</param>
+    /// <returns>Value of path</returns>
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:50]:
```suggestion
    /// <param name="expression">An expression. Typically the name of a variable or a column that contains JSON text.</param>
    /// <param name="path">A JSON path that specifies the object or the array to extract.</param>
    /// <returns>A JSON fragment.</returns>
    /// <seealso href="https://learn.microsoft.com/sql/t-sql/functions/json-query-transact-sql">SQL Server documentation for <c>JSON_QUERY</c>.</seealso>
```

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 7]
```diff
@@ -16,6 +16,44 @@ namespace Microsoft.EntityFrameworkCore;
...
+    public static string JsonQuery(
+        this DbFunctions _,
+        object propertyReference,
```
Comment (roji) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs:53]:
```suggestion
        object expression,
```

File: src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    /// </summary>
+    public SqlExpression? Translate(
+        SqlExpression? instance,
+        MethodInfo method,
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+    {
+        if (_methodInfoJsonFunctions.TryGetValue(method, out var function))
+        {
+            var propertyReference = arguments[1];
```
Comment (roji) [inline src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs:59]:
```suggestion
            var expression = arguments[1];
```

File: src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+            var typeMapping = propertyReference.TypeMapping;
+            var path = propertyReference.Type == arguments[2].Type
+                ? _sqlExpressionFactory.ApplyTypeMapping(arguments[2], typeMapping)
```
Comment (roji) [inline src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs:63]: This infers the type mapping of the path from the type mapping of the expression - I'm not sure that's right... These are two independent independent arguments, and I'm not sure the type mapping of e.g. the JSON column should affect the type mapping of the path.

So I'd just apply the default type mapping on the path in any case - unless you see a reason to do otherwise/

File: src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                : _sqlExpressionFactory.ApplyDefaultTypeMapping(arguments[2]);
+
+            var functionArguments = new List<SqlExpression> { propertyReference, path };
+
+            return _sqlExpressionFactory.Function(
+                function,
+                functionArguments,
+                nullable: true,
+                // TODO: don't propagate for now
```
Comment (roji) [inline src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs:72]: This needs to be implemented. From a quick test, a null expression propagates (i.e. JSON_VALUE returns null when its first argument is null), but a null path generates an error (so doesn't propagate).

File: src/EFCore.Sqlite.Core/Extensions/SqliteDbFunctionsExtensions.cs [hunk 11]
```diff
@@ -30,6 +30,23 @@ public static class SqliteDbFunctionsExtensions
...
+    ///     Maps to the SQLite <c>json_extract</c> function, which extracts and returns one or more values from well-formed JSON.
+    /// </summary>
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-database-functions">Database functions</see>, and
+    ///     <see href="https://aka.ms/efcore-docs-sqlite">Accessing SQLite databases with EF Core</see> for more information and examples.
+    /// </remarks>
+    /// <param name="_">The <see cref="DbFunctions" /> instance.</param>
+    /// <param name="propertyReference">The column</param>
+    /// /// <param name="paths">The column</param>
+    /// <returns>A hexadecimal string.</returns>
```
Comment (roji) [inline src/EFCore.Sqlite.Core/Extensions/SqliteDbFunctionsExtensions.cs:44]: Please fix these as for the SQL Server docs and add a `<seealso>` to the SQLite docs.

File: src/EFCore.Sqlite.Core/Extensions/SqliteDbFunctionsExtensions.cs [hunk 12]
```diff
@@ -30,6 +30,23 @@ public static class SqliteDbFunctionsExtensions
...
+    public static string JsonExtract(this DbFunctions _,
+        object propertyReference,
+        string paths)
```
Comment (roji) [inline src/EFCore.Sqlite.Core/Extensions/SqliteDbFunctionsExtensions.cs:47]: The docs say that json_extract accepts multiple path arguments (variadic), so this parameter seems wrong; shouldn't it be `params string[] paths`?

If so, the translator can throw if the array is non-constant.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    /// </summary>
+    public SqlExpression? Translate(
+        SqlExpression? instance,
+        MethodInfo method,
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+    {
+        if (_methodInfoJsonFunctions.TryGetValue(method, out var function))
+        {
+            var propertyReference = arguments[1];
```
Comment (roji) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:55]:
```suggestion
            var expression = arguments[1];
```

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+            var typeMapping = propertyReference.TypeMapping;
+
+            var path = propertyReference.Type == arguments[2].Type
+                        ? _sqlExpressionFactory.ApplyTypeMapping(arguments[2], typeMapping)
```
Comment (roji) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:60]: See comment above for SQL Server

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+                        : _sqlExpressionFactory.ApplyDefaultTypeMapping(arguments[2]);
+
+            var functionArguments = new List<SqlExpression> { propertyReference, path };
+
+            return _sqlExpressionFactory.Function(
+                function,
+                functionArguments,
+                nullable: true,
+                // TODO: don't propagate for now
```
Comment (roji) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:69]: As with SQL Server, you'll need to test what happens when you give NULL to json_extract in the various arguments and implement this accordingly.

**Review by roji — CHANGES_REQUESTED** [review]
Thanks for this PR - it's a good start, see comments below.

Note that there are some other JSON functions supported by SQL Server (ISJSON, JSON_ARRAY, JSON_OBJECT...). Ideally we'd have translations for all of them (and the relevant SQLite ones as well); of course you don't have to do that in this PR, but if you feel like working on those that would be great.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs (hunk 15 above)
Comment (joelmandell) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:81]: Thx for your comments @roji  :) I think I resolved all the other comments, except this one. You mean a test?

Comment (joelmandell) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:81]: Oh. So I did see some other comments now that I did not see previously. I will resolve those too

**Review by joelmandell — COMMENTED** [review]
.

**Review by joelmandell — COMMENTED** [review]
I made some changes now. And think I got them all covered. But little bit uncertain about the propagation of null.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        if (_methodInfoJsonFunctions.TryGetValue(method, out var function))
+        {
+            var expression = arguments[1];
+
+            var paths = arguments[2];
+
+            var functionArguments = new List<SqlExpression> { expression };
+
+            var pathValue = ((SqlConstantExpression)paths)?.Value;
```
Comment (roji) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:61]: See comment about non-constant path above

File: test/EFCore.Relational.Specification.Tests/TestModels/JsonQuery/JsonEntityBasicString.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.EntityFrameworkCore.TestModels.JsonQuery;
+public class JsonEntityBasicString
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/TestModels/JsonQuery/JsonEntityBasicString.cs:5]: Any reason to add this rather than simply using JsonEntityBasic?

File: test/EFCore.Sqlite.FunctionalTests/Query/JsonFunctionSqliteTest.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.EntityFrameworkCore.Sqlite.Internal;
+using Microsoft.EntityFrameworkCore.TestModels.JsonQuery;
+using Microsoft.EntityFrameworkCore.TestModels.JsonTest;
+
+namespace Microsoft.EntityFrameworkCore.Query;
+
+public class JsonFunctionSqliteTest
```
Comment (roji) [inline test/EFCore.Sqlite.FunctionalTests/Query/JsonFunctionSqliteTest.cs:10]: Note that we've recently merged support for JSON for SQLite, so you now have the same JsonQueryTestBase implementation as in SQL Server, so you should be able to add your tests to JsonQuerySqliteTest like to did above for SQL Server.

**Review by roji — CHANGES_REQUESTED** [review]
@joelmandell sorry, it's taking long for me to get around to this, here are some comments. Please rebase this PR on the latest main and check out any test failures after that.

File: src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+/// <summary>
+///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+public class SqlServerJsonFunctionsTranslator : IMethodCallTranslator
+{
+
```
[the remark this answers is not in the document]
Comment (joelmandell) [reply src/EFCore.SqlServer/Query/Internal/SqlServerJsonFunctionsTranslator.cs:16]: No problem. I will start looking into this PR later next week again because of other things in backlog 👍

**Comment by celsowm** [comment]
json_value in differents rdbms:

**mysql**: https://dbfiddle.uk/?rdbms=mysql_8.0&fiddle=591207ceddca6fe4f11a3ad8589e7c5b
**sqlserver**: https://dbfiddle.uk/?rdbms=sqlserver_2017&fiddle=5c70245f5bb7c96d4adcc6b10ebed195
**postgres**: https://dbfiddle.uk/o46x7dyL
**sqlite**: https://www.db-fiddle.com/f/tMcVDVmjxb5wSjemg34dy8/2

File: test/EFCore.Relational.Specification.Tests/TestModels/JsonQuery/JsonEntityBasicString.cs (hunk 17 above)
Comment (joelmandell) [reply test/EFCore.Relational.Specification.Tests/TestModels/JsonQuery/JsonEntityBasicString.cs:5]: JsonEntityBasic has mappings of the columns to classes, and this works only for strings If I get it correctly.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public SqlExpression? Translate(
+        SqlExpression? instance,
+        MethodInfo method,
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+    {
+        if(_methodInfoJsonFunctions.TryGetValue(method,out _))
```
Comment (Delizald) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteJsonFunctionsTranslator.cs:55]:
```suggestion
        if (_methodInfoJsonFunctions.TryGetValue(method,out _))
```

**Review by Delizald — CHANGES_REQUESTED** [review]

**Comment by joelmandell** [comment]
@roji I know I was late to the party and missed to tag you in to this before feature freeze for 8.0... Do you have any advice or comments on how I can move forward with this one to perhaps get it ready for 9.0? I can as well look into the other requests for json functions if possible.

**Comment by roji** [comment]
@joelmandell sure thing - I'd also definitely like us to get this in for 9.0.

Here's my advice... Give us around a month, maybe a bit more to deal with the 8.0 release and the subsequent patching and other work - and then ping me here again. Assuming it's a good time to iterate on it, you can then rebase this PR on the latest main, and then we can get this to completed state and merge.

**Review by Abdurahmon0412 — APPROVED** [review]

**Review by Abdurahmon0412 — CHANGES_REQUESTED** [review]

**Comment by joelmandell** [comment]
Ping @roji. I do not know if it is to early to continue on this one now? :)

**Comment by roji** [comment]
@joelmandell I'm looking at this PR again (sorry for all the breaks and pauses). Since there was previously some problematic merge commit which messed up the entire history, I've squashed everything cleanly rebased on top of the latest main. In the interest of finally getting moving quickly and finally getting this merge, I'll do whatever fixup is still necessary and push commits on this PR.

**Comment by joelmandell** [comment]
@roji Sounds good. Thx

**Comment by mseada94** [comment]
Having `JsonValue` extension method on `DbFunctions` may cause conflict when other provider wants to extend the same function,
For example if oracle provider want to map this function which has the same name and parameters, it should use different name otherwise it will conflict with SqlServer provider extension if user reference both providers.

The extension should be added on the relational level and the mapping is left to each provider to define.

**Comment by roji** [comment]
> Having JsonValue extension method on DbFunctions on sql server provider may cause conflict when other provider wants to extend the same function,
For example if oracle provider want to map this function which has the same name and parameters, it should use different name otherwise it will conflict with SqlServer provider extension if user reference both providers.
> The extension should be added on the relational level and the mapping is left to each provider to define.

You're probably right for this case.

Conflict between same-named extension methods across different providers is something we already have in various other places where multiple providers implement the same extension method (especially in model building). We generally tell users to simply use static method syntax instead of extension method - it seems much better for the few users who actually target mutiple databases to do this, rather than the vast majority of users who target a single database to have weird naming (FYI early versions of EF prefixed all extension methods with ForSqlServer and similar).

Putting something on the relational level usually has a high bar, since even when the function exists with the same name across different databases, there are typically slight differences in the signature etc.

However, for JSON_VALUE specifically (and probably some of the other JSON functions), it's part of the SQL standard and seems to be pretty closely followed across most databases which do support it:

* [SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/functions/json-value-transact-sql?view=sql-server-ver16) (no support for the RETURNING, ON EMPTY and ON ERROR clauses, but that's supposed to come)
* [PostgreSQL](https://www.postgresql.org/docs/current/functions-json.html)
* [Oracle](https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/JSON_VALUE.html)
* [MySQL](https://dev.mysql.com/doc/refman/8.4/en/json-search-functions.html#function_json-value)
* [MariaDB](https://mariadb.com/kb/en/json_value/) (no support for the RETURNING, ON EMPTY and ON ERROR clauses)
* SQLite - not supported (but they'll likely follow suit at some point, as they've done with the other PG JSON functions).

So I think this all makes sense.

Note the extra clauses RETURNING, ON EMPTY and ON ERROR which are widely supported (and probably in the standard). These have special syntax beyond the regular SQL function syntax, but we already internally have a special JsonScalarExpression which represents JSON_VALUE, and which could be augmented to support these clauses.

For the EF.Functions side, we could support these clauses via extra optional parameters - I'll see when implementing if that makes sense.

**Comment by roji** [comment]
> I think json expression should be tested before translation to allows only

In general, users using functions on EF.Functions take responsibility for what they do - it's a relatively direct way to specify a SQL function invocation, and there's not much point for us adding validation in EF if the database will error anyway. It's also possible for users to do some convoluted things, e.g. do JSON_VALUE over a non-column (parameter/literal) JSON document, with the JSON path coming from a database column; if we restricted the translation to only support the JSON type mapping, this scenario would be blocked (and again, for no good reason).

**Comment by mseada94** [comment]
> It's also possible for users to do some convoluted things, e.g. do JSON_VALUE over a non-column (parameter/literal) JSON document, with the JSON path coming from a database column

non-column (parameter/literal) JSON document should be of type string and the mapping should be of type `StringTypeMapping`

**Comment by roji** [comment]
@mseada94 right, so what's there to test exactly? What's needed beyond just typing the arguments as string, for now (until we make it possible to also invoke these functions with arbitrary objects mapped as complex types/owned entity types)?

**Comment by mseada94** [comment]
The test here to allow translation and calling for expression with type mapping (with first argument as object or generic type)
```cs
arguments[0].TypeMapping is SqlServerOwnedJsonTypeMapping or StringTypeMapping
```
- `StringTypeMapping`
  - string literal
  - string column
  - string parameter
  - string complex expression
  - arbitrary object configured to be mapped as string when storing on the database (value converter should be supported not json owned and complex types)
- `SqlServerOwnedJsonTypeMapping`
  - owned entity configured to be stored as json
- Future: complex type configured to be stored as json

It should prevent translation otherwise for example
- entity which is mapped to table
- owned entity which is mapped to table
- other datatypes (int, byte, Guid, Datetime, ....etc)

**Comment by mseada94** [comment]
Introducing complex type json support will reduce the need for using `ValueConverter` to store JSON, but it will not eliminate the need for it, so it should be supported as well (Dictionary) could be converted and might not be supported by Complex Type, and some users might need to have custom json serialization and deserialization and can't use complex type.

---

please consider the following example 
- using `JsonValue` function with Address is valid, the value is serialized and converted to json (That's the reason to have first argument as object or generic not only string) 
- However, Department here is a table not a json and the translation should not be allowed here (That's the reason for having validation)

```cs
using Microsoft.EntityFrameworkCore;
using System.Text.Json;

using var context = new SomeDbContext();

var validUsage = context.Users.Where(u => EF.Functions.JsonValue(u.Address, "$.Country") == "Egypt");
var inValidUsage = context.Users.Where(u => EF.Functions.JsonValue(u.Department, "$.Name") == "Engineering");

public class User
{
    public Guid Id { get; set; }
    public Address Address { get; set; }
    public Department Department { get; set; }

}

public class Department
{
    public Guid Id { get; set; }
    public string Name { get; set; }
}


public class Address
{
    public string Country { get; set; }
    public string City { get; set; }
}


public class SomeDbContext : DbContext
{
    public DbSet<User> Users { get; set; }
    public DbSet<Department> Departments { get; set; }
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder.UseSqlServer(Your.ConnectionString);

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<User>().HasOne(d => d.Department).WithMany();

        modelBuilder.Entity<User>().Property(x => x.Address).HasConversion(
            address => JsonSerializer.Serialize(address, DefaultJsonSerializerOptions),
            json => JsonSerializer.Deserialize<Address>(json, DefaultJsonSerializerOptions)!);
    }

    private static JsonSerializerOptions DefaultJsonSerializerOptions = new JsonSerializerOptions
    {
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };
}
```

File: src/EFCore.Sqlite.Core/Extensions/SqliteDbFunctionsExtensions.cs [hunk 21]
```diff
@@ -30,6 +30,36 @@ public static class SqliteDbFunctionsExtensions
...
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-database-functions">Database functions</see>, and
+    ///     <see href="https://aka.ms/efcore-docs-sqlite">Accessing SQLite databases with EF Core</see> for more information and examples.
+    /// </remarks>
+    /// <param name="_">The <see cref="DbFunctions" /> instance.</param>
+    /// <param name="expression">The json</param>
+    /// <param name="path">The paths</param>
+    /// <returns>INTEGER or REAL for a JSON numeric value, an INTEGER zero for a JSON false value, an INTEGER one for a JSON true value, the dequoted text for a JSON string value, and a text representation for JSON object and array values</returns>
+    /// <seealso href="https://www.sqlite.org/json1.html#the_json_extract_function">SQLite documentation for <c>json_extract</c>.</seealso>
+    public static object JsonExtract(this DbFunctions _, object expression, string path)
```
Comment (ljani) [inline src/EFCore.Sqlite.Core/Extensions/SqliteDbFunctionsExtensions.cs:60]: Thanks for the great feature. I tested this in my project where `path` is a(n interpolated) variable and got the following exception:

```
System.InvalidCastException: 'Unable to cast object of type 'Microsoft.EntityFrameworkCore.Query.SqlExpressions.SqlParameterExpression' to type 'Microsoft.EntityFrameworkCore.Query.SqlExpressions.SqlConstantExpression'.'
```

(I forgot to copy the stacktrace, but it was from [`SqliteJsonFunctionsTranslator.cs#60`](https://github.com/dotnet/efcore/pull/30010/files#diff-59b487e208d8edc777c63571006756baf22c7473799a44c6d74b91461fa40c86R60) with the `SqlConstantExpression` cast.)

This was fixed by tagging `path` with `[NotParameterized]` here. I'm not 100 % sure if this is the correct fix, but something to consider.

**Comment by roji** [comment]
@joelmandell FYI I've just submitted #37477, which adds EF.Functions.JsonExists to relational (a bit similar to what you were doing here for other JSON access functions).

As lots of time has passed (mainly our fault), please let me know if you're still interested in carrying on work in this PR for EF 11 - it would be good to deliver a pack of JSON translations together as a holistic feature. If so:

* Please take a look at the testing infrastructure introduced in #37477 - that's the ideal place for testing to be in this PR as well.
* I'm also interested in possibly including EF.Functions method stubs in relational rather than as different stubs in each provider - similar to how [#37477](https://github.com/dotnet/efcore/pull/37477/changes) does for JsonExists. Specifically, JSON_VALUE and JSON_QUERY are included in the SQL standard, even though they may not be universally-implemented (e.g. they don't exist on SQLite). In other words, it may make sense for us to introduce standard JsonValue/JsonQuery on EF.Functions stubs in relational, and have each provider translate them to whatever makes sense. Let me know if you have any thoughts on this.

Again, sorry this has stalled for so long, but I'd be interested in delivering it for EF 11.

**Comment by joelmandell** [comment]
No worries @roji 
I will look into it
